include/ibamr/vc_ins_utilities.h

```diff
@@ -0,0 +1,411 @@
+
+    /*!
+     * Set the viscosity based on the current level set information for three-phase flows.
+     */
+    void setViscosityPatchData3PhaseFlows(int mu_idx,
+                                          SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > mu_var,
+                                          SAMRAI::tbox::Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                                          const int cycle_num,
+                                          const double time,
+                                          const double current_time,
+                                          const double new_time);
+
+    /*!
+     * Name of this object.
+     */
+    std::string d_object_name;
+
+    /*!
+     * Pointer to advection-diffusion solver.
+     */
+    SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> d_adv_diff_solver;
+
+    /*!
+     * Level set variable.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > d_ls_gas_var, d_ls_solid_var;
+
+    /*!
+     * Density.
+     */
+    double d_rho_liquid, d_rho_gas, d_rho_solid;
+
+    /*!
+     * Viscosity.
+     */
+    double d_mu_liquid, d_mu_gas, d_mu_solid;
+
+    /*!
+     * Number of interface cells over which to smooth the material properties.
+     */
+    double d_num_gas_interface_cells, d_num_solid_interface_cells;
+
+    /*!
+     * Whether or not to set viscosity in the solid region.
+     */
+    bool d_set_mu_solid;
+
+    /*!
+     * Number of phases. Valid options are 2 and 3.
+     */
+    int d_num_phases = 2;
+
+}; // SetFluidProperties
+
+/*!
+ * \brief The GravityForcing class provides an implementation of gravity force.
+ * This class can be utilized to set the gravity force \f$ \rho g \f$ throughout the entire domain with
+ * the option grav_type = "FULL" or specifically within the liquid and gas regions (excluding the solid phase)
+ * with the option grav_type = "FLOW".
+ */
+class GravityForcing : public IBTK::CartGridFunction
+{
+public:
+    /*!
+     * \brief Constructor for this class.
+     *
+     * grav_const stores the acceleration due to gravity.
+     *
+     */
+    GravityForcing(const std::string& object_name,
+                   SAMRAI::tbox::Pointer<INSVCStaggeredHierarchyIntegrator> ins_hierarchy_integrator,
+                   std::vector<double> grav_const);
+
+    /*!
+     * \brief Constructor for this class.
+     *
+     * @param input_db provides parameters such as rho_neg, rho_pos, and num_gas_interface_cells.
+     * grav_const stores the acceleration due to gravity.
+     *
+     */
+    GravityForcing(const std::string& object_name,
+                   SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_hierarchy_integrator,
+                   SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_gas_var,
+                   SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                   std::vector<double> grav_const);
```

Comment:
Make sure this is correct. Where is the gravity force specifically applied? How are `rho_neg` and `rho_pos` used?
```suggestion
    /*!
     * \brief Constructor for this class. Applies the gravitational force to only the portion of domain where the level set is negative.
     *
     * @param input_db provides parameters such as rho_neg, rho_pos, and num_gas_interface_cells.
     * grav_const stores the acceleration due to gravity.
     *
     */
    GravityForcing(const std::string& object_name,
                   SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_hierarchy_integrator,
                   SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_gas_var,
                   SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
                   std::vector<double> grav_const);
```

Reply:
I have verified this. The gravity force is applied across the entire domain. However, the density used in the gravity term only accounts for the densities of the liquid and gas phases, excluding the solid density. Specifically, rho_pos and rho_neg represent the densities of the liquid and gas phases, respectively.